auth-backends/src/main/scala/bootstrap/rudder/plugin/AuthBackendsConf.scala

```
@@ -1367,15 +1375,82 @@ object RudderOpaqueTokenAuthenticationProvider {
 
+  /*
+   * The cache semantic is:
+   * - if the key is not defined, we don't have that token in cache: IdP validation is needed
+   * - if the key is defined:
+   *   - when `Left(ex)`, it means that the token is not valid and we can fail the authentication
+   *   - when `Right(auth)`, it means we can check if the auth is already ok.
+   */
+  private val cache = validationCache match {
+    // negative or too small duration are the same as no cache defined.
+    case Some(duration) if (duration.toMillis > 0) =>
+      Some(
+        Caffeine
+          .newBuilder()
+          .maximumSize(1_000)
+          .expireAfterWrite(java.time.Duration.ofMillis(duration.toMillis))
+          .build[String, Either[AuthenticationException, RudderOAuth2OpaqueToken]]()
+      )
+
+    case _ => None
+  }
+
+  /*
+   * The (pure) method to use in case of cache miss.
+   */
+  private def doRemoteAuthentication(
+      authentication: BearerTokenAuthenticationToken
+  ): Either[AuthenticationException, RudderOAuth2OpaqueToken] = {
+    try {
+      opaqueTokenAuthenticationProvider.authenticate(authentication) match {
+        case token: RudderOAuth2OpaqueToken => Right(token)
+
+        case _ => Left(new InvalidBearerTokenException(s"Token is not of the expected RudderOAuth2OpaqueToken type"))
+      }
+    } catch {
+      case ex: AuthenticationException =>
+        Left(ex)
+    }
+  }
+
   override def authenticate(authentication: Authentication): Authentication = {
-    val a = opaqueTokenAuthenticationProvider.authenticate(authentication)
-    a
+    authentication match {
+      case token: BearerTokenAuthenticationToken =>
+        // check is that token already have an authentication
+        cache match {
+          // cache defined:
+          case Some(c) =>
+            c.get(token.getToken, (_: String) => doRemoteAuthentication(token)) match {
+              case Left(ex)                                      => throw ex
```

is seems like `Left` values errors are always stored in cache. Is it not potentially an issue if the called is filled up with wrong tokens even so that valid tokens are discarded ?
Why wouldn't we want to simply return the error ?

There seems to be solutions to discard the value in the cache : like returning null or so...
But it's really a Java API, we could use https://github.com/blemale/scaffeine